libcxx/include/__iterator/static_bounded_iter.h

```diff
@@ -0,0 +1,268 @@
+    __current_ -= __n;
+    return *this;
+  }
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX14 friend __static_bounded_iter
+  operator-(__static_bounded_iter const& __self, difference_type __n) _NOEXCEPT {
+    __static_bounded_iter __tmp(__self);
+    __tmp -= __n;
+    return __tmp;
+  }
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX14 friend difference_type
+  operator-(__static_bounded_iter const& __x, __static_bounded_iter const& __y) _NOEXCEPT {
+    return __x.__current_ - __y.__current_;
+  }
+
+  // Comparison operations.
+  //
+  // These operations do not check whether the iterators are within their bounds.
+  // The valid range for each iterator is also not considered as part of the comparison,
+  // i.e. two iterators pointing to the same location will be considered equal even
+  // if they have different validity ranges.
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR friend bool
+  operator==(__static_bounded_iter const& __x, __static_bounded_iter const& __y) _NOEXCEPT {
+    return __x.__current_ == __y.__current_;
+  }
+
+#if _LIBCPP_STD_VER <= 17
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR friend bool
+  operator!=(__static_bounded_iter const& __x, __static_bounded_iter const& __y) _NOEXCEPT {
+    return __x.__current_ != __y.__current_;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR friend bool
+  operator<(__static_bounded_iter const& __x, __static_bounded_iter const& __y) _NOEXCEPT {
+    return __x.__current_ < __y.__current_;
+  }
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR friend bool
+  operator>(__static_bounded_iter const& __x, __static_bounded_iter const& __y) _NOEXCEPT {
+    return __x.__current_ > __y.__current_;
+  }
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR friend bool
+  operator<=(__static_bounded_iter const& __x, __static_bounded_iter const& __y) _NOEXCEPT {
+    return __x.__current_ <= __y.__current_;
+  }
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR friend bool
+  operator>=(__static_bounded_iter const& __x, __static_bounded_iter const& __y) _NOEXCEPT {
+    return __x.__current_ >= __y.__current_;
+  }
+
+#else
+  _LIBCPP_HIDE_FROM_ABI constexpr friend strong_ordering
+  operator<=>(__static_bounded_iter const& __x, __static_bounded_iter const& __y) noexcept {
+    if constexpr (three_way_comparable<_Iterator, strong_ordering>) {
+      return __x.__current_ <=> __y.__current_;
+    } else {
+      if (__x.__current_ < __y.__current_)
+        return strong_ordering::less;
+
+      if (__x.__current_ == __y.__current_)
+        return strong_ordering::equal;
+
+      return strong_ordering::greater;
+    }
+  }
+#endif // _LIBCPP_STD_VER >= 20
+
+private:
+  template <class>
+  friend struct pointer_traits;
+  template <class, size_t, class>
+  friend struct __static_bounded_iter;
+  _Iterator __current_; // current iterator
+  _Iterator __begin_;   // start of the valid range, which is [__begin_, __begin_ + _Size)
```

Comment:
We could make this a pointer + size pair instead, which would make the iterator a bit smaller. It could also help the compiler with optimizations, since we're always just using an offset from the same pointer. Not sure though. Is there a reason we're doing it this way instead?

Note: It would also be possible to store the offset in the low bits of the pointer in some cases (e.g. `array<std::string, 7>`), making it even smaller.

Reply:
100% agreed that this needs to be investigated. This is actually the target of https://github.com/llvm/llvm-project/issues/108600. My thinking is that we'd start by shipping this, and then we could investigate the best representation for both `__static_bounded_iter` and `__bounded_iter`, since both should use the same representation.

About storing the offset in the low bits of the pointer, it's true that we could do that in some cases, but the cost of unpacking the pointer, doing the increment and re-packing the pointer would likely not be worth it? After all, these iterators are expected to be used mostly as local variables, not stored in data structures all over the place. So IMO we should optimize for their modifying operations to be as cheap as possible. WDYT?

Reply:
If it's just a local variable I think the compiler should be able to unpack the bits once, modify it however it wants and re-pack at the end. I think we agree that most use cases are `std::some_algo(arr.begin(), arr.end())`? I'm pretty sure the compiler would be able to completely resolve the bit-packing and just use the pointer. We'd have to investigate though.

Reply:
With my latest implementation, the representation of the pointer + bounds are extracted into their own utility. I think it would be reasonably easy to swap out for a different representation. I'd suggest we land the basic functionality + test changes, and then investigate these representation optimizations separately for both the static and the non-static variant of bounded iterator.